src/ripple/server/impl/Role.cpp

```diff
@@ -21,72 +21,46 @@
 
 namespace ripple {
 
-Role
-adminRole (HTTP::Port const& port, Json::Value const& params,
-    beast::IP::Endpoint const& remoteIp,
-        std::vector<beast::IP::Endpoint> const& admin_allow)
-{
-    Role role (Role::FORBID);
-
-    bool const bPasswordSupplied =
-        params.isMember ("admin_user") ||
-        params.isMember ("admin_password");
-
-    bool const bPasswordRequired =
-        ! port.admin_user.empty() || ! port.admin_password.empty();
-
-    bool bPasswordWrong;
-
-    if (bPasswordSupplied)
-    {
-        if (bPasswordRequired)
-        {
-            // Required, and supplied, check match
-            bPasswordWrong =
-                (port.admin_user !=
-                    (params.isMember ("admin_user") ? params["admin_user"].asString () : ""))
-                ||
-                (port.admin_password !=
-                    (params.isMember ("admin_user") ? params["admin_password"].asString () : ""));
-        }
-        else
-        {
-            // Not required, but supplied
-            bPasswordWrong = false;
-        }
-    }
-    else
-    {
-        // Required but not supplied,
-        bPasswordWrong = bPasswordRequired;
-    }
-
-    // Meets IP restriction for admin.
-    beast::IP::Endpoint const remote_addr (remoteIp.at_port (0));
-    bool bAdminIP = false;
-
-    // VFALCO TODO Don't use this!
-    for (auto const& allow_addr : admin_allow)
-    {
-        if (allow_addr == remote_addr)
-        {
-            bAdminIP = true;
-            break;
-        }
-    }
+bool
+passwordUnrequiredOrSentCorrect (HTTP::Port const& port,
+                                 Json::Value const& params) {
 
-    if (bPasswordWrong                          // Wrong
-            || (bPasswordSupplied && !bAdminIP))    // Supplied and doesn't meet IP filter.
-    {
-        role   = Role::FORBID;
-    }
-    // If supplied, password is correct.
-    else
-    {
-        // Allow admin, if from admin IP and no password is required or it was supplied and correct.
-        role = bAdminIP && (!bPasswordRequired || bPasswordSupplied) ? Role::ADMIN : Role::GUEST;
-    }
+    assert(port.allow_admin);
+    bool const passwordRequired = (!port.admin_user.empty() ||
+                                   !port.admin_password.empty());
 
+    return !passwordRequired  ||
+            ((params["admin_password"].isString() &&
```

Comment:
Might was well just say: `if (params["admin_password"].asString() == port.admin_password)` because it's going to be exceedingly rare when it is _not_ a string, and `asString()` will do the right thing in that case.

(Same comment as below...)